Add int1 to int7 dtypes

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #136301

Summary:
Similar to https://github.com/pytorch/pytorch/pull/117208, we want to add int1 to int7 for edge use cases
for weight quantization (https://www.internalfb.com/diff/D62464487)

Test Plan:
python test/test_quantization.py -k test_uint4_int4_dtype

Reviewers:

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Add int1 to int7 dtypes #117208

Summary:
Similar to https://github.com/pytorch/pytorch/pull/117208, we want to add int1 to int7 for edge use cases
for weight quantization

Test Plan:
python test/test_quantization.py -k test_uint4_int4_dtype

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "Add int1 to int7 dtypes"


Summary:
Similar to https://github.com/pytorch/pytorch/pull/117208, we want to add int1 to int7 for edge use cases
for weight quantization

Test Plan:
python test/test_quantization.py -k test_uint4_int4_dtype

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

ezyang commented:
Can you explain in more detail why the signed types are desirable?

jerryzh168 commented:
@ezyang it's used in edge low bit kernels to represent weight: https://www.internalfb.com/diff/D62464487

TBH, I think it we can extend torch.dtype out of core, it's probably better, but my impression is that we can't do that right now

jerryzh168 committed:
Update on "Add int1 to int7 dtypes"


Summary:
Similar to https://github.com/pytorch/pytorch/pull/117208, we want to add int1 to int7 for edge use cases
for weight quantization (https://www.internalfb.com/diff/D62464487)

Test Plan:
python test/test_quantization.py -k test_uint4_int4_dtype

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
actually we can delay this PR I think, since we have been uint8 right now for uintx as well: https://github.com/pytorch/ao/blob/53b6b787bb035f1abc3c4804dd22c09ee41e7a0d/torchao/dtypes/uintx/uintx.py#L66, we just need to add some placeholder dtypes in torchao in this case.

jerryzh168 commented:
@ezyang you mean just for int1 or other intx dtypes as well? I agree int1 is probably not very useful. but cc @metascroy and also @kimishpatel to see if there are more context here for the need intx dtypes (especially int1).

ezyang commented:
For the larger int types, it seems to me that the biggest benefit you get over unsigned is access to signed operations. But this also isn't obviously that helpful; for example, in two's complement, you can implemented signed addition... as regular unsigned addition.

Obviously, exposing signed integers as a user facing abstraction is a legitimate thing to do, but my understanding of AO's use case is they would wrapped up in a tensor subclass anyway...

kimishpatel commented on c10/core/ScalarType.h:
```diff
@@ -467,14 +479,22 @@ inline bool isSignedType(ScalarType t) {
     case ScalarType::UInt5:
     case ScalarType::UInt6:
     case ScalarType::UInt7:
-      return true;
```
This seems like a bug before this pr?

jerryzh168 replied:
yeah I think so

kimishpatel commented on test/quantization/core/test_utils.py:
```diff
@@ -192,30 +192,31 @@ def test_quantize_weight_clamping_per_channel(self):
         assert quantized_tensor.int_repr().min().item() == q8_min
 
-    def test_uint1_7_dtype(self):
+    def test_uint4_int4_dtype(self):
```
what happened to the original test case?

jerryzh168 replied:
naming is a bit off, original test is testing uint4, and I added int4 in the PR

kimishpatel reviewed (commented):
left some comments

kimishpatel commented:
@ezyang so in general it is not clear why unsigned is treated different than signed? We have had this issue with uint16. So just from UX perspective it seems like an aberration that we have uintx types but not intx.

> but my understanding of AO's use case is they would wrapped up in a tensor subclass anyway

I think this is fair and I am not sure for the kernels we actually need intx/uintx types at all if kernels are tied with quantization API. Meaning if we did not use AO API, and relied purely on exported IR to derive if some tensors are uintx/intx, then I think having this natively available in PyTorch makes sense.

But in general @ezyang what would be the objection given uintx are already present.

jerryzh168 commented:
> I think this is fair and I am not sure for the kernels we actually need intx/uintx types at all if kernels are tied with quantization API. Meaning if we did not use AO API, and relied purely on exported IR to derive if some tensors are uintx/intx, then I think having this natively available in PyTorch makes sense.

yes all the operations are implemented with tensor subclass (currently just pack/unpack to/from uint8), we are using torch.uintx/torch.intx as user facing API so people don't need to maybe use (nbits + signed bool) to express them

jerryzh168 committed:
Update on "Add int1 to int7 dtypes"


Summary:
Similar to https://github.com/pytorch/pytorch/pull/117208, we want to add int1 to int7 for edge use cases
for weight quantization (https://www.internalfb.com/diff/D62464487)

Test Plan:
python test/test_quantization.py -k test_uint4_int4_dtype

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@ezyang any ideas on the CI failure (https://github.com/pytorch/pytorch/actions/runs/11026420870/job/30727936288?pr=136301)? I can't repro locally

ezyang commented:
looks like master problem

jerryzh168 commented:
@pytorchbot rebase

jerryzh168 committed:
Update

[ghstack-poisoned]

jerryzh168 commented:
@pytorchbot merge

PaliC commented:
@pytorchbot revert -m "causing internal failures" -c "ghfirst"

jerryzh168 commented:
landed in https://github.com/pytorch/pytorch/pull/137928
